File: bot/src/services/soil.py
import logging

import httpx

logger = logging.getLogger(__name__)

ISRIC_SOILGRIDS_URL = "https://rest.isric.org/soilgrids/v2.0/classification/query"

WRB_RISK_MAP = {
    "Acrisol": 0.7,
    "Albeluvisol": 0.5,
    "Alisol": 0.7,
    "Andosol": 0.3,
    "Arenosol": 0.6,
    "Cambisol": 0.4,
    "Chernozem": 0.2,
    "Ferralsol": 0.5,
    "Fluvisol": 0.8,
    "Gleysol": 0.9,
    "Histosol": 0.9,
    "Kastanozem": 0.3,
    "Leptosol": 0.7,
    "Luvisol": 0.4,
    "Nitisol": 0.3,
    "Phaeozem": 0.3,
    "Planosol": 0.7,
    "Plinthosol": 0.6,
    "Podzol": 0.5,
    "Regosol": 0.6,
    "Solonchak": 0.8,
    "Solonetz": 0.7,
    "Umbrisol": 0.4,
    "Vertisol": 0.9,
}

WRB_NAME_MAP = {
    "Acrisol": "Acrisol (tanah masam)",
    "Albeluvisol": "Albeluvisol",
    "Alisol": "Alisol (tanah masam)",
    "Andosol": "Andosol (tanah vulkanik)",
    "Arenosol": "Arenosol (tanah berpasir)",
    "Cambisol": "Cambisol (tanah muda)",
    "Chernozem": "Chernozem (tanah hitam)",
    "Ferralsol": "Ferralsol (tanah merah)",
    "Fluvisol": "Fluvisol (tanah endapan)",
    "Gleysol": "Gleysol (tanah jenuh air)",
    "Histosol": "Histosol (tanah gambut)",
    "Kastanozem": "Kastanozem",
    "Leptosol": "Leptosol (tanah tipis)",
    "Luvisol": "Luvisol (tanah liat)",
    "Nitisol": "Nitisol (tanah subur)",
    "Phaeozem": "Phaeozem",
    "Planosol": "Planosol",
    "Plinthosol": "Plinthosol",
    "Podzol": "Podzol (tanah pasir)",
    "Regosol": "Regosol (tanah longgar)",
    "Solonchak": "Solonchak (tanah asin)",
    "Solonetz": "Solonetz (tanah soda)",
    "Umbrisol": "Umbrisol",
    "Vertisol": "Vertisol (tanah lempung)",
}


async def get_soil_type(lat: float, lon: float, timeout: float = 8.0) -> dict | None:
    params = {"lat": lat, "lon": lon}

    try:
        async with httpx.AsyncClient(timeout=timeout) as client:
            resp = await client.get(ISRIC_SOILGRIDS_URL, params=params)
            resp.raise_for_status()
            data = resp.json()

            wrb = data.get("wrb_class_name")
            if not wrb or wrb == "No information":
                logger.info("No soil data at (%.4f, %.4f)", lat, lon)
                return None

            risk_score = WRB_RISK_MAP.get(wrb, 0.5)
            name_id = WRB_NAME_MAP.get(wrb, wrb)

            result = {
                "code": wrb,
                "name": name_id,
                "risk_score": risk_score,
            }

            logger.info(
                "Soil at (%.4f, %.4f): %s (risk=%.2f)",
                lat, lon, wrb, risk_score,
            )
            return result

    except Exception as e:
        logger.warning("Soil fetch failed for (%.4f, %.4f): %s", lat, lon, e)
        return None

File: web-app/src/lib/risk/soil.ts
const ISRIC_API = 'https://rest.isric.org/soilgrids/v2.0/classification/query';
const TIMEOUT_MS = 5000;

interface SoilInfo {
  score: number;
  name: string;
}

const WRB_MAP: Record<string, { score: number; name: string }> = {
  Acrisol: { score: 0.7, name: 'Acrisol (tanah masam)' },
  Albeluvisol: { score: 0.5, name: 'Albeluvisol' },
  Alisol: { score: 0.7, name: 'Alisol (tanah masam)' },
  Andosol: { score: 0.3, name: 'Andosol (tanah vulkanik)' },
  Arenosol: { score: 0.6, name: 'Arenosol (tanah berpasir)' },
  Cambisol: { score: 0.4, name: 'Cambisol (tanah muda)' },
  Chernozem: { score: 0.2, name: 'Chernozem (tanah hitam)' },
  Ferralsol: { score: 0.5, name: 'Ferralsol (tanah merah)' },
  Fluvisol: { score: 0.8, name: 'Fluvisol (tanah endapan)' },
  Gleysol: { score: 0.9, name: 'Gleysol (tanah jenuh air)' },
  Histosol: { score: 0.9, name: 'Histosol (tanah gambut)' },
  Kastanozem: { score: 0.3, name: 'Kastanozem' },
  Leptosol: { score: 0.7, name: 'Leptosol (tanah tipis)' },
  Luvisol: { score: 0.4, name: 'Luvisol (tanah liat)' },
  Nitisol: { score: 0.3, name: 'Nitisol (tanah subur)' },
  Phaeozem: { score: 0.3, name: 'Phaeozem' },
  Planosol: { score: 0.7, name: 'Planosol' },
  Plinthosol: { score: 0.6, name: 'Plinthosol' },
  Podzol: { score: 0.5, name: 'Podzol (tanah pasir)' },
  Regosol: { score: 0.6, name: 'Regosol (tanah longgar)' },
  Solonchak: { score: 0.8, name: 'Solonchak (tanah asin)' },
  Solonetz: { score: 0.7, name: 'Solonetz (tanah soda)' },
  Umbrisol: { score: 0.4, name: 'Umbrisol' },
  Vertisol: { score: 0.9, name: 'Vertisol (tanah lempung)' },
};

const cache = new Map<string, SoilInfo>();

function cacheKey(lat: number, lon: number): string {
  const k = (v: number) => Math.round(v * 1e3) / 1e3;
  return `${k(lat)},${k(lon)}`;
}

export async function getSoilType(
  latitude: number,
  longitude: number,
): Promise<SoilInfo | null> {
  const key = cacheKey(latitude, longitude);
  const cached = cache.get(key);
  if (cached) return cached;

  try {
    const url = `${ISRIC_API}?lat=${latitude}&lon=${longitude}`;
    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), TIMEOUT_MS);

    const res = await fetch(url, { signal: controller.signal });
    clearTimeout(timeout);

    if (!res.ok) return null;

    const json = await res.json() as { wrb_class_name?: string };
    const wrb = json.wrb_class_name;
    if (!wrb || wrb === 'No information') return null;

    const mapped = WRB_MAP[wrb] ?? { score: 0.5, name: wrb };
    const result: SoilInfo = { score: mapped.score, name: mapped.name };
    cache.set(key, result);
    return result;
  } catch {
    return null;
  }
}
